104.md

```diff
@@ -0,0 +1,205 @@
+
+## New Event Kinds
+
+This NIP introduces three new event kinds. First, we'll look at the two that are related to the conversation itself. These event kinds MUST always be gift-wrapped as per [NIP-59](https://github.com/nostr-protocol/nips/blob/master/59.md) and use [NIP-44](https://github.com/nostr-protocol/nips/blob/master/44.md) encryption for the [Seal (kind 13)](https://github.com/nostr-protocol/nips/blob/master/59.md#2-the-seal-event-kind) and [Gift Wrap (kind 1059)](https://github.com/nostr-protocol/nips/blob/master/59.md#3-gift-wrap-event-kind) events.
+
+This combination might seem like overkill given we're further encrypting the actual content of the messages between the users using a double-ratchet symmetrical encryption scheme but this combination significantly decreases the amount of metadata published and improves deniability/repudiation of the conversation.
+
+### Conversation Request Event
+
+This event is sent when one user wants to start a conversation with another user. For example, if Alice wants to talk to Bob, her first message will be a conversation request event. This message allows both parties to establish a shared root key via a Diffie-Hellman key exchange and allows Alice to derive the first key in the chain to encrypt a message. An encrypted message MUST be included in the conversation request event to ensure that the first ratchet step is performed. The message can be a default message like, "Hi, I would like to connect." if no message is provided by the user.
+
+```json
+{
+    "kind": 443,
+    "created_at": <unix timestamp in seconds>,
+    "pubkey": <pubkey of the sender>,
+    "content": <ciphertext of initial message symmetrically encyrpted using first sending chain/message key>,
+    "tags": [
+        ["p", <pubkey of the recipient>],
+        ["prekey", <pubkey of the recipient prekey used for DH calculation>],
+        ["ephemeral", <pubkey of the ephemeral key used for DH calculation>]
+    ],
+    "sig": "" // No signature because these are only sent inside gift wraps.
+}
+```
+
+### Double Ratchet Direct Message (DRDM) Event
+
+The Double Ratchet Direct Message (DRDM) event is the main event kind that carries the conversation content. This event carries enough information for recievers to properly double ratchet their chains and decrypt the message content with the right symmetrical keys.
+
+```json
+{
+    "kind": 444,
+    "created_at": <unix timestamp in seconds>,
+    "pubkey": <pubkey of the sender>,
+    "content": <ciphertext of the message symmetrically encrypted with chain/message key>,
+    "tags": [
+        ["p", <pubkey of the recipient>],
+        ["dh_sending", <current DH sending pubkey>],
+        ["current_index", <number of current message in the current message chain>],
+        ["previous_length", <length of previous message chain>]
```

Comment:
When sharing the key between clients/devices, do we also need to share the current rachet state (`current_index`, `previous_length`? Or should a new client start from 0?

Reply:
This DOES NOT work out of the box on multiple clients or devices. Each client/device combo should be thought of as a different "inbox" that will have it's own initial setup and will maintain it's own set of chains/keys for each conversation with another person. At a very high level, say I have Primal on iOS and Amethyst on Android and you have just Amethyst on Android. You're client would have to know that it's having the same conversation with two device-clients but using different sets of keys/chains. It would encrypt each message you send for each recipient and then publish those GW events separately.

Reply:
This is similar to how this sort of scheme works for group conversations as well.

Reply:
@staab @paulmillr high level description of multiple devices question here... ☝️  

you can also read more about how signal does it here: https://www.signal.org/docs/specifications/sesame/

I want to read up a bit more before committing to any path here. In an ideal world we have multi-device conversations, it's significantly more complex for clients. From the beginning I viewed this more as a solution that a dedicated secure messaging client would be using, and less as the normal standard for DMs in social clients. That said, nothing stopping those clients from implementing it, just a question of making the UI good enough.

Reply:
If that is the case, then I would add a: 

Clients MUST not re-use/re-import keys from other clients or devices. The Key is not supposed to leave the app/device it was created from.

Reply:
I agree. I need to make this whole device/client limitation more clear in the NIP.